packages/apollo-server-core/src/caching.ts

```diff
@@ -1,4 +1,67 @@
+import { ExecutionResult } from 'graphql';
+
 export interface PersistedQueryCache {
   set(key: string, data: string): Promise<any>;
   get(key: string): Promise<string | null>;
 }
+
+export function calculateCacheControlHeaders(
+  responses: Array<ExecutionResult & { extensions?: Record<string, any> }>,
+) {
+  let maxAge = Number.MAX_VALUE;
+  let publicOrPrivate = 'public';
+
+  //Because of the early exit, we are unable to use forEach. While a reduce
+  //loop might be possible, a for loop is more readable
+  for (let i = 0; i < responses.length; i++) {
+    const response = responses[i];
+
+    const cacheControl: {
+      version: number;
+      hints: Array<{ scope?: string; maxAge?: number; path: Array<string> }>;
+    } =
+      response.extensions && response.extensions.cacheControl;
+
+    //If there are no extensions or hints, then the headers should not be present
+    if (
+      !cacheControl ||
+      !cacheControl.hints ||
+      cacheControl.hints.length === 0
+    ) {
+      return {};
+    }
+
+    const rootHints = new Set<string>();
+    for (let y = 0; y < cacheControl.hints.length; y++) {
+      if (cacheControl.hints[y].scope === 'PRIVATE') {
+        publicOrPrivate = 'private';
+      }
+
+      //If no maxAge is present, then we ignore the hint
+      if (cacheControl.hints[y].maxAge === undefined) {
+        continue;
+      }
+
+      //if there is a hint with max age of 0, we don't need to process more
+      if (cacheControl.hints[y].maxAge === 0) {
+        return {};
+      }
+
+      if (cacheControl.hints[y].maxAge < maxAge) {
+        maxAge = cacheControl.hints[y].maxAge;
+      }
+
+      rootHints.add(cacheControl.hints[y].path[0]);
```

Comment:
This is not quite right (or at least, not quite what we did before).  This would mean that if you got a hint for "foo.bar" but nothing else, we would count this as being a hint on "foo".  But if there's also a "foo.baz" and that was the only hint, then there are no hints on "foo.baz", so it wouldn't be right to count this as covering everything.

(Arguably we could change the "all roots must have hints" logic to be something more like "every leaf must be the descendent of a node with a hint" although that wouldn't be quite right either, as even interior nodes have resolvers whose return values can change.)